modules/ping/check.go

```diff
@@ -0,0 +1,315 @@
+package ping
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+	"net"
+	"time"
+
+	"github.com/wtfutil/wtf/logger"
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+)
+
+const (
+	ProtocolICMP     = 1
+	ProtocolIPv6ICMP = 58
+	msgFail          = "fail"
+	msgSuccess       = "success"
+)
+
+func checkIPV4(target string, pingTimeout int, logging bool) (result string) {
+	dst, err := net.ResolveIPAddr("ip4", target)
+	if err != nil {
+		return msgFail
+	}
+
+	var conn *icmp.PacketConn
+	conn, err = icmp.ListenPacket("ip4:icmp", "0.0.0.0")
```

Comment:
This is why it requires root, right? Could this use `udp4` instead and get around the root requirement? I would prefer for people not to have to run as root if possible.

Reply:
I read that you need to [tweak kernel settings for Linux](https://sturmflut.github.io/linux/ubuntu/2015/01/17/unprivileged-icmp-sockets-on-linux/) and didn't see anything obvious for MacOS. Again, for selfish purposes, I'm happy to run wtf as root as I understand the code I'm running but I fully understand the dangers of running as root and that your average user may not, so will have a look at a non-root alternative. There may be times when TCP trumps UDP, and vice versa, so it may be UDP becomes the default and TCP optional, but with a warning over kernel tweaking/root privileges. A weekend project awaits :)

Reply:
This is a cool module. If it can be made to work non-root, I'd happily include it in.

Reply:
It seems there are a couple of ways to handle this for Linux (without setuid or running as root):
1) Set file capabilities on the wtf binary so that the ping module can use the socket, e.g.
`setcap cap_net_raw=eip <binary>`  
Result: Works as expected

2) Enable user to use ICMP_ECHO sockets (a feature added to the Linux kernel to avoid root/setuid), e.g. 
`sysctl -w "net.ipv4.ping_group_range=0 65535"`
Result: Didn't work. Suspect the failure relates to recently reported issue here: https://github.com/golang/go/issues/39891

Will take a look at the udp4 suggestion.

Reply:
For what it's worth, I wouldn't consider running wtfutil as root.

Reply:
Has shelling out to ping been considered? The parsing pain is a downside, but it completely bypasses the permissions issue of creating low-level sockets (and the permissions problem is supposed to be hard to get around, because "security"). `ping --c1 IP` would even bypass any parsing, it returns 0 for sucess, non-zero for failure. That seems pretty easy.
```
% ping -c1 no.such.domain; echo $?; ping -c1 example.com; echo $? 
ping: no.such.domain: Name or service not known
2
PING example.com(2606:2800:220:1:248:1893:25c8:1946 (2606:2800:220:1:248:1893:25c8:1946)) 56 data bytes
64 bytes from 2606:2800:220:1:248:1893:25c8:1946 (2606:2800:220:1:248:1893:25c8:1946): icmp_seq=1 ttl=59 time=20.1 ms

--- example.com ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 20.143/20.143/20.143/0.000 ms
0
```

Reply:
@acaloiaro I've removed the root requirement. Linux users would need to change the kernel config as mentioned above.
@sam-github Aside from parsing, the lack of concurrency could make it very slow if running multiple pings. Unless you can fork a number of concurrent ping processes?
I've spent days trying to work out a concurrency bug and, IIRC, I can now reproduce consistently. It works fine on my Raspberry PI, and an AWS VM, but not on my Macbook when using Wifi. I _suspect_ it's an issue with Go and Wifi, or certain network adapters (works fine on Macbook with network cable) and have written a small piece of code to help prove to the Go community. I now need to tidy it up and see if anyone else can help.
Happy for this to be closed and I'll submit a new PR when I've got to the bottom of this.

Reply:
Forking multiple ping processes is like forking multiple of any other kind of process, no problem at all.

Reply:
Maybe that's the way to go then. I'll still try and get this working (for my own sanity at least!) but will close whilst there might be a better solution plus I've little time right now.

Reply:
If you want to keep this PR open (or re-open it), feel free. This is a good conversation and it would be cool to have this in WTF. Up to you.